[CEP XXXX] OCI Storage of Conda Artifacts (#115)
conda/ceps: cep-XXXX.md

<tr><td> Created </td><td> April 12, 2024</td></tr>
<tr><td> Updated </td><td> March 10, 2025</td></tr>
<tr><td> Discussion </td><td> https://github.com/conda/ceps/pull/115; https://github.com/conda/ceps/pull/70 </td></tr>
<tr><td> Implementation </td><td>  </td></tr>
</table>

## Abstract

This CEP defines how to store conda packages in an *OCI registry* as *OCI artifacts* (see the [Open Containers Initiative](https://opencontainers.org/) for more details on these terms). Importantly, this CEP does NOT describe how to store conda repodata in an OCI registry or how to generate repodata from an OCI registry. We leave these topics for future CEPs.

## Specification

The conda OCI package specification has three main parts. First, we define how the data in a conda package is stored as an OCI artifact. Second, we codify the allowed characters and format of conda channels, subdirs, labels, and package names. Finally, we define how to generate the container `<name>` and container image `<tag>` (i.e., `<name>:<tag>`) for a conda package in an OCI registry.

Below we use the following terms, some of which are defined by the OCI specs and others of which are specific to conda.

> OCI-define terms:

Reviewer: ```suggestion
OCI-defined terms:
```

> - *OCI registry*: A service that implements the OCI Distribution Spec (e.g., `ghcr.io`).

Reviewer: Ah you linked below, nvm.

> - *OCI repository*: A collection of container images with the same repository `<name>` in an OCI registry.

Reviewer: Ok, no, the repository is actually the OCI image or container name. Confusing but I guess that's their official terminology. I would have understood it better like this:

```suggestion
- *OCI repository*: A collection of OCI artifacts with the container or image `<name>` in an OCI registry.
```

This specification is labeled `v1` and uses only `v1.0` of the OCI specification.

### conda Packages as OCI Artifacts

In order to store a conda package as an OCI artifact, we need to define how the data within the conda package is mapped into the OCI blobs, the allowed blob mediaTypes, and what additional metadata (if any) is stored in the OCI manifest.

#### conda OIC Artifact Blob/Layer Media Types

We define the following custom media types that MUST be used for the storage of conda packages in an OCI registry.

| Blob type        | Content type              | mediaType                                   |
|------------------|---------------------------|---------------------------------------------|
| conda package v1 | .tar.bz2 package          | application/vnd.conda.package.v1            |
| conda package v2 | .conda package            | application/vnd.conda.package.v2            |
| package info     | `info` folder as gzip     | application/vnd.conda.info.v1.tar+gzip      |

> | package index    | `index.json` file         | application/vnd.conda.info.index.v1+json    |

Reviewer: ```suggestion
| package index    | `info/index.json` file         | application/vnd.conda.info.index.v1+json    |
```

#### conda OCI Artifact Blob/Layer Structure

A valid conda package as an OCI artifact MUST have the following layers:

> - a layer holding either the `.tar.bz2` or `.conda` package file with the mediaType `application/vnd.conda.package.v1` or `application/vnd.conda.package.v2`, respectively.

Reviewer: Some packages (especially in defaults) are packaged as both .tar.bz2 and .conda. Can't those channels include both layers in the same artifact?

Author: No they cannot. conda index, the last I checked, assumed those packages hold the same information, so we should pick one and we should prefer the `.conda` one. I have added updated text to the spec.

Reviewer: Yes, but they have different SHA's. With `.conda` we need to assume ZSTD decompressors are available, but I guess that's a given now.

- a layer holding the `info` folder as a gzipped tarball with extension `.tar.gz` and mediaType `application/vnd.conda.info.v1.tar+gzip`.

> - a layer holding the `index.json` file as plain JSON with the mediaType `application/vnd.conda.info.index.v1+json`.

Reviewer: ```suggestion
- a layer holding the `info/index.json` file as plain JSON with the mediaType `application/vnd.conda.info.index.v1+json`.
```

Additional layers are NOT allowed in the OCI artifact.

#### conda Manifest Annotations

The manifest MUST have the following [Annotations](https://github.com/opencontainers/image-spec/blob/main/annotations.md#annotations):

- `org.conda.oci.schema` with the value `1` indicating the `v1` specification (defined by this document).
- `org.conda.package.name` with the name of the conda package.
- `org.conda.package.version` with the version of the conda package.
- `org.conda.package.build` with the build string of the conda package

> Additional annotations are NOT allowed.

Reviewer: I think we cannot guarantee this because providers like GHCR might add their own too? At any case, we can say something like:

```suggestion
Additional annotations under the `org.conda` namespace are NOT allowed.
```

### Allowed Characters and Formats for conda Channels, Subdirs, Labels, and Package Names

> The follow regexes define valid conda channel names, labels, and conda package names:

Reviewer: I assume this only standardizes names for OCI, not the whole ecosystem?

Reviewer: Yes, I think these need to be put in a separate CEP (which can be added later or in parallel) where we standardize the naming schemes for all the components in the ecosystem.

Reviewer: Opened https://github.com/conda/ceps/pull/116

- channel: `^[a-z0-9]+((-|_|.)[a-z0-9]+)*$`
- subdirs: `^[a-z0-9]+((-|_|.)[a-z0-9]+)*$`
- label: `^[a-zA-Z][0-9a-zA-Z_\-\.\/:\s]*`
- package name: `^(([a-z0-9])|([a-z0-9_](?!_)))[._-]?([a-z0-9]+(\.|-|_|$))*$`

> All channels, subdirs, labels, and package names MUST conform to their respective regex in the list above regardless of whether or not they are package inside an OCI registry or a tradition conda package index (e.g., anaconda.org).

Author: ```suggestion
All channels, subdirs, labels, and package names MUST conform to their respective regex in the list above regardless of whether or not they are package inside an OCI registry or a traditional conda package index (e.g., anaconda.org).
```

Further, the following rule applies to labels:

- The label `NOLABEL` is reserved and MUST only be used for conda packages which have no other labels. In other words, in the space of labels, the empty set is represented by the label `NOLABEL`.

### Mapping conda Packages with Channels, Subdirs, Labels to OCI Repositories and Tags

Per the [OCI Distribution Spec](https://github.com/opencontainers/distribution-spec) v1.0, the `<name>` and `<tag>` of an OCI repository are subject to the following regexes:

- `<name>`: `^[a-z0-9]+((\.|_|__|-+)[a-z0-9]+)*(\/[a-z0-9]+((\.|_|__|-+)[a-z0-9]+)*)*$`
- `<tag>`: `^[a-zA-Z0-9_][a-zA-Z0-9._-]{0,127}$`

The [OCI Distribution Spec](https://github.com/opencontainers/distribution-spec) further notes that "Many clients impose a limit of 255 characters on the length of the concatenation of the registry hostname (and optional port), /, and <name> value."

> These constraints motivate the following set of rules for mapping conda packages to OCI repositories and tags. More explanation is given below in the Rational section and the example implementation illustrate how these rules work together to create a human-readable and easily parsable OCI repository `<name>` and `<tag>` for a conda package.

Reviewer: ```suggestion
These constraints motivate the following set of rules for mapping conda packages to OCI repositories and tags. More explanation is given below in the Rationale section and the example implementation illustrate how these rules work together to create a human-readable and easily parsable OCI repository `<name>` and `<tag>` for a conda package.
```

#### Overall Form of the OCI Repository and Tag

A conda package with its channel, subdir, and (optional) label MUST be mapped to an OCI `<repository>:<tag>` in one of two forms.

The unhashed form of the OCI repository and tag for a conda package is:

```
<channel>/<subdir>/<OCI-encoded package name>:<OCI-encoded version, build and (optional) label in the form '<version>-<build>(-<label>)'>
```

In the hashed form, the OCI-form of the package name and the string composed of the OCI-forms of the version, build, and (optional) label (i.e., `<version>-<build>(-<label>)`) are each separately hashed via SHA1 and then replaced by the string `h<hexdigest of SHA1 hash>` as follows

```
<channel>/<subdir>/h<hexdigest of the SHA1 hash of OCI-encoded package name>:h<hexdigest of the SHA1 hash of the original OCI tag of the conda package>
```

The `<channel>` and `<subdir>` MUST be used as-is without modification. The OCI-forms of the package name, label, version, and build string are defined below.

#### Encoding Package Names, Channels, and Subdirs into OCI Repository `<name>`s

The package name MUST be encoded to OCI-form as follows:

> - If the package name starts with an `_`, the `_` MUST be replaced with `z`.

Reviewer: Why replaced and not prepended?

Author: To keep the names as short as possible. We could instead just prepend everything with a `c` at the cost of a single extra character.

Author: @jaimergp thoughts here?

Reviewer: I'm not sure if a single character is worth the trouble of having a rule for this, specially given how you've assessed that we are far off from the limit in general. But no strong feelings against leaving things as they are written right now.

> - Otherwise, if the package name does NOT start with an `_`, the package name MUST be prepended by `c`.

Reviewer: Can you elaborate on why these characters were chosen? I assume `z` is just to sort last, and `c` for `conda` just to have something else?

Author: I was following the suggestion from @baszalmstra.

> - If the combined string `<channel>/<subdir>/<OCI-encoded package name>` exceeds 128 characters in length, the OCI-encoded package name MUST be replaced by the SHA1 hash of the OCI-encoded package name in hexadecimal form as `h<SHA1 in hexidecimal>`.

Reviewer: ```suggestion
- If the combined string `<channel>/<subdir>/<OCI-encoded package name>` exceeds 128 characters in length, the OCI-encoded package name MUST be replaced by the SHA1 hash of the OCI-encoded package name in hexadecimal form as `h<SHA1 in hexadecimal>`.
```

Package names in OCI-form which are not hashed can be decoded by reversing the encoding rules above.

For example, the package name `_libgcc_mutex` is encoded to OCI-form as `zlibgcc_mutex`, and the package name `zlibgcc_mutex` is encoded to OCI-form as `czlibgcc_mutex`. A very long package name like `gblah000...000` would be encoded to its OCI-form `cgblah000...000`, then hashed via SHA1, yielding something like `h<hexdigest of SHA1 hash>`.

The channel and subdir MUST be used as-is without modification.

The channel, subdir and OCI-encoded package name (or hashed OCI-encoded package name if required) MUST be combined with `/` to form the OCI repository `<name>`

```
<channel>/<subdir>/<OCI-encoded package name>
```

#### Encoding Labels, Versions and Build Strings into OCI `<tag>`s

The label, version, and build string are encoded to OCI-form as follows. First the following mapping rules MUST be applied to the label, version and build string in the order listed below:

> - `_` -> `__`

Reviewer: Only single underscores are allowed as input, right? Because OCI only allows two consecutive underscores IIRC.

Author: Yep. Changed to `_U` per your suggestion!

- `-` -> `_D`
- `+` -> `_P`
- `!` -> `_N`
- `=` -> `_E`
- `:` -> `_C`
- `/` -> `_S`
- ` ` -> `_B`
- `\t` -> `_T`
- `\r` -> `_R`

> - `\n` -> `_L`

Reviewer: Since only two consecutive underscores are allowed, what happens with an input like `something_+else`? It would become `something___Pelse`, which has three underscores. Maybe `_` needs to become `_U`?

Author: you are correct! fixed!

```
v1.0.0-h123456__0-main
v1.0.0-h123456__0
```

Finally, if the entire OCI tag exceeds 128 characters in length, the entire tag MUST be hashed with the SHA1 algorithm and the resulting OCI tag is then `h<SHA1 in hexidecimal>`.

#### Further Odds and Ends

- If either the OCI-encoded form of package name or the OCI tag for the conda package are hashed, then both MUST be hashed. In other words, only the two forms of the OCI artifact `<name>:<tag>` combination defined above, either unhashed or hashed, are allowed.
- Once the OCI repository `<name>` and `<tag>` are computed for the conda artifact, the rest of the OCI Distribution Spec MUST be followed.
- If an OCI registry implementation errors due to length when dealing with the full OCI form of the conda package (e.g., the combination of the registry URL and full OCI repository `<name>` for the conda package exceed some internal limit in the implementation), then that package cannot be stored in that registry and an error MUST be raised.

> - Within conda repodata, the download URL for a conda package in an OCI registry MUST always be prepended by `oci://` to indicate that the download URL is an OCI URL. The registry being used, any required ports etc. are prepended to the conda package OCI repository `<name>:<tag>` to form the full download URL per the OCI v1.0 specification.

Reviewer: I think this topic might require further details, maybe in the upcoming OCI-compatible repodata CEP?

Author: I want to include this clause here so that we can start including these urls in existing repodata files even if the repodata focused spec does not exist yet. Is there anything unclear here?

Reviewer: I think we need to be a bit more specific then. Something like:

> The download URL OCI-mirrored packages MUST use the `oci://` scheme. More specifically, it MUST follow the syntax `oci://<authority>[:port]/[path]/<name>:<tag>`.

Note that `<name>:<tag>` in the path component of the URL is technically illegal; or the colon needs to be %-encoded. We could also add it as one more path component (`.../<name>/<tag>`) or a qualifier (`.../<name>?tag=<tag>`). 

The thing is that I can't find an actual spec in the OpenContainers org where they recommend `oci://` as the scheme, but there are some real world examples of apps having come up with that convention, like `helm` or `kustomizer`.

Author: Yeah these are not real urls. They are instead what you feed to the OCI runtime utility like the docker command line tool.

#### Overall Form of the OCI Repository and Tag

A conda package with its channel, subdir, and (optional) label MUST be mapped to an OCI `<repository>:<tag>` in one of two forms.

The unhashed form of the OCI artifact `<name>:<tag>` for a conda package is:

```
<channel>/<subdir>/<OCI-encoded package name>:<OCI-encoded version>-<OCI-encoded build>(-<OCI-encoded label>)
```

where the `-<OCI-encoded label>` component MUST be present if the package label is not `main`.

In the hashed form, the OCI-form of the package name and the OCI tag (i.e., `<OCI-encoded version>-<OCI-encoded build>(-<OCI-encoded label>)`) are each separately hashed via SHA1 and then replaced by the string `h<hexdigest of SHA1 hash>` as follows:

```
<channel>/<subdir>/h<hexdigest of the SHA1 hash of OCI-encoded package name>:h<hexdigest of the SHA1 hash of the original OCI tag of the conda package>
```

> The `<channel>` and `<subdir>` MUST be used as-is without modification. The OCI-forms of the package name, label, version, and build string are defined below.

Reviewer: What if the channel name along is too long to be encoded? Would it be ok to have top-level `h<hash>` entries with the corresponding `org.conda.channel.{name,subdir,label}` annotation metadata?

Author: Yes maybe just doing that all the time is simpler actually. I was trying to increase readability, but the complexity cost is high.

- *OCI manifest*: A JSON document that describes the contents of an OCI artifact, organized as some metadata and a number of blobs. The manifest is also stored in the registry as a blob.

Conda-specific terms:

- *OCI-encoded package name*: The name of a conda package that is encoded to a specific format (defined below) for use in an OCI repository `<name>`. The language "encoded to OCI-form" is used below to refer to this process.
- *OCI-encoded label/version/build string*: A conda package label, version, or build string that is encoded to a specific format (defined below) for use in an OCI image `<tag>`. The language "encoded to OCI-form" is used below to refer to this process.
- *hashed OCI-encoded <item>*: An `<item>` that is first encoded to its OCI-form and then hashed to a specific format (defined below) for use in an OCI repository `<name>` and/or `<tag>`.

For further details on the OCI-defined terms, see the full OCI specifications, the [Image Spec](https://github.com/opencontainers/image-spec) and the [Distribution Spec](https://github.com/opencontainers/distribution-spec).

> This specification is labeled `v1` and uses only `v1.0` of the OCI specification.

Author: ```suggestion
This specification is labeled `v1` and uses only OCI specifications compatible with the OCI `v1` specification.
```

- `org.conda.package.version` with the version of the conda package.
- `org.conda.package.build` with the build string of the conda package

Additional annotations under the `org.conda` namespace are NOT allowed.

### Allowed Characters and Formats for conda Channels, Subdirs, Labels, and Package Names

The follow regexes define valid conda channel names, labels, and conda package names for conda packages that can be stored in an OCI registry.

- channel: `^[a-z0-9]+((-|_|.)[a-z0-9]+)*$`
- subdirs: `^[a-z0-9]+((-|_|.)[a-z0-9]+)*$`
- label: `^[a-zA-Z][0-9a-zA-Z_\-\.\/:\s]*`
- package name: `^(([a-z0-9])|([a-z0-9_](?!_)))[._-]?([a-z0-9]+(\.|-|_|$))*$`

All channels, subdirs, labels, and package names MUST conform to their respective regex in the list above if the package is to be stored in an OCI registry.

Further, the following rule applies to labels:

- The label `NOLABEL` is reserved and MUST only be used for conda packages which have no other labels. In other words, in the space of labels, the empty set is represented by the label `NOLABEL`.

### Mapping conda Packages with Channels, Subdirs, Labels to OCI Repositories and Tags

> Per the [OCI Distribution Spec](https://github.com/opencontainers/distribution-spec) v1.0, the `<name>` and `<tag>` of an OCI repository are subject to the following regexes:

Author: ```suggestion
Per the [OCI Distribution Spec](https://github.com/opencontainers/distribution-spec) v1, the `<name>` and `<tag>` of an OCI repository are subject to the following regexes:
```

The channel and subdir MUST be used as-is without modification.

The channel, subdir and OCI-encoded package name (or hashed OCI-encoded package name if required) MUST be combined with `/` to form the OCI repository `<name>`

```
<channel>/<subdir>/<OCI-encoded package name or its SHA1 hash in hexadecimal>
```

#### Encoding Labels, Versions and Build Strings into OCI `<tag>`s

The label, version, and build string are encoded to OCI-form as follows. First the following mapping rules MUST be applied to the label, version and build string in the order listed below:

- `_` -> `_U`
- `-` -> `_D`
- `+` -> `_P`
- `!` -> `_N`
- `=` -> `_E`
- `:` -> `_C`
- `/` -> `_S`
- ` ` -> `_B`
- `\t` -> `_T`
- `\r` -> `_R`
- `\n` -> `_L`

This encoding can be undone by applying the rules in reverse, starting at the bottom of the list and moving to the top. Depending on the context, some labels may be percent-encoded for use in URLs. The percent-encoding MUST be undone before the label is encoded via the list of rules above.

> After encoding each component to OCI-form, the version, build string, and (optional) label MUST be combined with a `-` to form the OCI `<tag>` as either

Author: ```suggestion
After encoding each component to OCI-form, the version, build string, and label MUST be combined with a `-` to form the OCI `<tag>`
```

```

> <OCI-encoded version>-<OCI-encoded build>-<OCI-encoded label>

Author: ```suggestion
<OCI-encoded version>-<OCI-encoded build>(-<OCI-encoded label>)
```

```

> or if the label is not specified (i.e. the package is on the `main` label), then

Author: ```suggestion
where the `-<OCI-encoded label> component is present is the package label is not `main`.
```

> ```
> <OCI-encoded version>-<OCI-encoded build>
> ```
> 
> with no trailing `-`.
> 

Author: ```suggestion
```

## Alternatives

- An initial version of this specification was proposed in [CEP PR #70](https://github.com/conda/ceps/pull/70) and received substantial feedback from the community. The current version of the specification is a result of that feedback. Particularly notable feedback included the suggestion of [symbol name mangling techniques and the first version of the shortened encoding used above](https://github.com/conda/ceps/pull/70/files#discussion_r1740438239) by [Bas Zalmstra](https://github.com/baszalmstra).

- The [channel-mirrors project](https://github.com/channel-mirrors) with its associated package [conda-oci-mirror](https://github.com/channel-mirrors/conda-oci-mirror) developed a proof-of-concept implementation of conda packages and repodata stored in an OCI registry. This specification is based on the experience gained from that project, but is not fully backwards compatible with the original implementation. See Backwards Compatibility above. This original implementation and implied specification is referred to as `v0`.

## Sample Implementation

### conda-oci

> The [conda-oci](https://github.com/conda-incubator/conda-oci) repository has reference implementations of the procedures defined above for computing the OCI artifact `<name>:<tag>` and for converting a conda artifact into an OCI artifact with a manifest and layers/blobs.

Reviewer: This is not ready yet, right?

Author: Nope. My plan is to have it ready before the vote happens. The code parsing/validation is done, but I need to add in the package structure, CI, and code to convert artifacts.

The package name MUST be encoded to OCI-form using the following rules

- All package names MUST be prepended by the character `c`.
- If the combined string `<channel>/<subdir>/<OCI-encoded package name>` exceeds 128 characters in length, the OCI-encoded package name MUST be replaced by the SHA1 hash of the OCI-encoded package name in hexadecimal form as `h<SHA1 in hexadecimal>`. The letter `h` is reserved to indicate a hashed OCI-encoded package name.

Package names in OCI-form which are not hashed can be decoded by reversing the encoding rules above.

For example, the package name `_libgcc_mutex` is encoded to OCI-form as `c_libgcc_mutex`, and the package name `c_libgcc_mutex` is encoded to OCI-form as `cc_zlibgcc_mutex`. A very long package name like `gblah000...000` would be encoded to its OCI-form `cgblah000...000`, then hashed via SHA1, yielding something like `h<hexdigest of SHA1 hash>`.

The channel and subdir MUST be used as-is without modification.

The channel, subdir and OCI-encoded package name (or hashed OCI-encoded package name if required) MUST be combined with `/` to form the OCI repository `<name>`

```

> <channel>/<subdir>/<OCI-encoded package name>

Author: ```suggestion
<channel>/<subdir>/<OCI-encoded package name or its SHA1 hash in hexidecimal>
```

```

#### Encoding Labels, Versions and Build Strings into OCI `<tag>`s

The label, version, and build string are encoded to OCI-form as follows. First the following mapping rules MUST be applied to the label, version and build string in the order listed below:

- `_` -> `_U`
- `-` -> `_D`
- `+` -> `_P`
- `!` -> `_N`
- `=` -> `_E`
- `:` -> `_C`
- `/` -> `_S`
- ` ` -> `_B`
- `\t` -> `_T`
- `\r` -> `_R`
- `\n` -> `_L`

This encoding can be undone by applying the rules in reverse, starting at the bottom of the list and moving to the top. Depending on the context, some labels may be percent-encoded for use in URLs. The percent-encoding MUST be undone before the label is encoded via the list of rules above.

After encoding each component to OCI-form, the version, build string, and (optional) label MUST be combined with a `-` to form the OCI `<tag>` as either

```

> <version>-<build>-<label>

Author: ```suggestion
<OCI-encoded version>-<OCI-encoded build>-<OCI-encoded label>
```

```

or if the label is not specified (i.e. the package is on the `main` label), then

```

> <version>-<build>

Author: ```suggestion
<OCI-encoded version>-<OCI-encoded build>
```

```

with no trailing `-`.

As implied above, if no label is explicitly specified for a package, then package is by definition on the `main` label. This stipulation means the following two OCI tags for conda packages are equivalent:

```

> v1.0.0-h123456__0-main
> v1.0.0-h123456__0

Author: ```suggestion
v1.0.0-h123456_U0-main
v1.0.0-h123456_U0
```

```

Finally, if the entire OCI tag exceeds 128 characters in length, the entire tag MUST be hashed with the SHA1 algorithm and the resulting OCI tag is then `h<SHA1 in hexadecimal>`.

#### Further Odds and Ends

- If either the OCI-encoded form of package name or the OCI tag for the conda package are hashed, then both MUST be hashed. In other words, only the two forms of the OCI artifact `<name>:<tag>` combination defined above, either unhashed or hashed, are allowed.
- Once the OCI repository `<name>` and `<tag>` are computed for the conda artifact, the rest of the OCI Distribution Spec MUST be followed.
- If an OCI registry implementation errors due to length when dealing with the full OCI form of the conda package (e.g., the combination of the registry URL and full OCI repository `<name>` for the conda package exceed some internal limit in the implementation), then that package cannot be stored in that registry and an error MUST be raised.

> - The download URL for a conda package in an OCI registry MUST use the `oci://` scheme to indicate that the artifact will be downloaded from an OCI endpoint. More specifically, it MUST follow the syntax `oci://<authority>[:<port>]/<name>:<tag>`. The `<authority>` is the URI of the OCI registry (e.g., `ghcr.io`).

Author: ```suggestion
- The repodata download URL for a conda package in an OCI registry MUST use the `oci://` scheme to indicate that the artifact will be downloaded from an OCI endpoint. More specifically, it MUST follow the syntax `oci://<authority>[:<port>]/<name>:<tag>`. The `<authority>` is the URI of the OCI registry (e.g., `ghcr.io`).
```

## Rationale

The set of rules defined above ensure that

- Nearly all conda packages can be stored in an OCI registry without hashing. As of 2025-03-10, the maximum lengths across `defaults` and `conda-forge` for the various components are ~90 characters, well below the OCI limits.
- conda packages whose OCI artifact `<name>:<tag>` are not hashed retain human-readability.
- For conda packages whose OCI artifact `<name>:<tag>` are not hashed, the underlying conda package information can be decoded from the OCI artifact `<name>:<tag>` without needing to access the OCI registry or consult a lookup table.
- For conda packages whose OCI artifact `<name>:<tag>` are hashed, the underlying conda package information can be extracted from the OCI Annotations stored in the OCI manifest.
- All labels on `anaconda.org` can be used as-is.
- Relabeling a conda package in an OCI registry is a cheap operation since it involves retagging an existing image.

Some specific choices were made to ease parsing and avoid edge cases:

- The encoding of labels, versions, and build strings is purposefully short to avoid excessively increasing the length of OCI tags.

> - The encoding of labels, versions, and build strings only generates at most two `_`s in a row and so obeys the OCI spec.

Author: ```suggestion
```

This string will typically be `<hostname>[:<port>]` (e.g., `ghcr.io` with no port specified). OCI conda channel URLs MUST
use `oci` as the scheme. OCI-compatible channel path and labels are defined below. If the `<OCI-encoded label>` is `main`,
the entire optional label component MUST be omitted. Otherwise, the label component MUST be present.

The base URL for an OCI conda channel is the canonical, unique identifier for the channel itself and SHOULD be used in
conda ecosystem tools to refer to the channel.

Conda `<OCI-compatible subdir>`s are appended to this base URL as follows:

```text
<OCI conda channel base URL>/<OCI-compatible subdir>
```

where `<OCI-compatible subdir>` is defined below.

The following regexes MUST be satisfied by the `<OCI-compatible channel path>`, `<OCI-compatible subdir>`, and `<OCI-compatible label>`:

- `<OCI-compatible channel path>`: `^[a-z0-9]+((\.|_|__|-+)[a-z0-9]+)*(\/[a-z0-9]+((\.|_|__|-+)[a-z0-9]+)*)*$` (same as the regex for an OCI repository `<name>`)
- `<OCI-compatible label>`: `^[a-z0-9]+((\.|_|__|-+)[a-z0-9]+)*(\/[a-z0-9]+((\.|_|__|-+)[a-z0-9]+)*)*$` (same as the regex for an OCI repository `<name>`)

> - `<OCI-compatible subdir>`: `^[a-z0-9]+(-[a-z0-9]+)*$`

Reviewer: This is the same as the one proposed in https://github.com/conda/ceps/pull/116, right? We should update this part to refer to that one once accepted.

### Subdir Repository Namespace Designations

The namespace after the combination `<OCI conda channel base URL>/<OCI-compatible subdir>/` has the following designations
(where the `*` wildcard indicates any valid OCI repository `<name>` component or part of a component as allowed by
the OCI repository `<name>` regex):

- `<OCI conda channel base URL>/<OCI-compatible subdir>/r*`: reserved for repodata JSON artifacts, run exports JSON artifacts, and
  future CEP-defined extensions for sharded repodata (defined in [CEP 16][cep16]).
- `<OCI conda channel base URL>/<OCI-compatible subdir>/c*`: reserved for OCI-encoded conda package names (The character `c`
  is prepended to the conda package name to indicate that it is an OCI-encoded conda package name as described below.)
- `<OCI conda channel base URL>/<OCI-compatible subdir>/h*`: reserved for hashed OCI-encoded conda package names
- `<OCI conda channel base URL>/<OCI-compatible subdir>/p*`: reserved for repodata patch instructions

> These designations are used to prevent namespace collisions between the various components of an OCI conda channel.

Reviewer: Should we add a clause that prevents any other initial letters from being used altogether? Something like:

```suggestion
These designations are used to prevent namespace collisions between the various components of an OCI conda channel. Any other string prefixes MUST NOT be used.
```

### Conda Artifact OCI Repository Names

Conda artifacts MUST be stored in the OCI repository of an OCI conda channel under their `<OCI-compatible subdir>` as the OCI repository

```text
<OCI conda channel base URL>/<OCI-compatible subdir>/<OCI-encoded package name>
```

The contents and mediaTypes of OCI blobs that make up the conda package artifact, along with how to compute the OCI tag, are defined below as well.

The `<OCI-encoded package name>` MUST be computed according to the following rules:

- All package names MUST be prepended by the character `c`.

> - If the combined string `<OCI conda channel base URL>/<OCI-compatible subdir>/<OCI-encoded package name>` exceeds 128 characters

Author: ```suggestion
- If the `<OCI-encoded package name>` exceeds 64 characters
```

  in length, the`<OCI-encoded package name>` MUST be replaced by the SHA256 hash of the `<OCI-encoded package name>`
  in hexadecimal form as `h<SHA256 in hexadecimal>`. The letter `h` is reserved to indicate a hashed OCI-encoded package name as described above.

`<OCI-encoded package name>`s that are not hashed can be decoded by reversing the encoding rules above.

For example, the package name `_libgcc_mutex` is encoded to OCI-form as `c_libgcc_mutex`, and the package name `c_libgcc_mutex`
is encoded to OCI-form as `cc_zlibgcc_mutex`. A very long package name like `gblah000...000` would be encoded to its OCI-form

> `cgblah000...000`, then hashed via SHA256, yielding something like `h<hexdigest of SHA256 hash>`.

Reviewer: Let's add a note here saying that the original components can be found as annotation metadata, as discussed below.

Author: So to be clear, I didn't end up hashing everything like we discussed. It is just the package name that is hashed. The rationale here is two-fold

1. Packages move between channels and so it didn't make sense to me to hash the channel and label with the package name. 

2. The hashes would occupy a specific namespace in the space of channels. While collisions themselves through normal operations would be rare, I was concerned that there might be security implications for this down the road.

### Conda Artifact OCI Tags

The OCI tag for a conda package MUST be computed from the packages `<OCI-encoded version>` and `<OCI-encoded build string>` as follows.

First, the `<OCI-encoded version>` and `<OCI-encoded build string>` are computed from conda package's `<version>` and `<build string>`
respectively. To do this, one MUST apply the mapping rules below to the conda package's `<version>` or `<build string>`
in the order listed, top to bottom:

> - `_` -> `__`

Reviewer: We are reverting the `_U` thing from last iteration, right? Just making sure it's not an oversight because in tag names we did find that multiple sequential underscores are ok.

Reviewer: Ah yes, you mentioned this below.

### Allowed Blob mediaTypes

We define the following custom media types that MUST be used for the storage of conda repodata, run exports, channel data,
and packages in an OCI registry:

| Blob type            | Content type                | mediaType                                      |
|----------------------|-----------------------------|------------------------------------------------|
| conda package v1     | .tar.bz2 package            | application/vnd.conda.package.v1               |
| conda package v2     | .conda package              | application/vnd.conda.package.v2               |
| package info         | `info` folder as gzip       | application/vnd.conda.info.v1.tar+gzip         |
| package index        | `info/index.json` file      | application/vnd.conda.info.index.v1+json       |
| repodata.json        | `repodata.json` file        | application/vnd.conda.repodata.v1+json         |
| repodata.json.zst    | `repodata.json.zst` file    | application/vnd.conda.repodata.v1+json+zst     |
| repodata.json.gz     | `repodata.json.gz` file     | application/vnd.conda.repodata.v1+json+gzip    |

> | repodata.json.bz2    | `repodata.json.bz2` file    | application/vnd.conda.repodata.v1+json+bz2     |

Reviewer: Do we need to add `v2` repodatas (CEP 15)?

Author: Yes

| channeldata.json     | `channeldata.json` file     | application/vnd.conda.channeldata.v1+json      |
| run_exports.json     | `run_exports.json` file     | application/vnd.conda.run_exports.v1+json      |
| run_exports.json.zst | `run_exports.json.zst` file | application/vnd.conda.run_exports.v1+json+zst  |
| run_exports.json.gz  | `run_exports.json.gz` file  | application/vnd.conda.run_exports.v1+json+gzip |
| run_exports.json.bz2 | `run_exports.json.bz2` file | application/vnd.conda.run_exports.v1+json+bz2  |
| repodata_from_packages.json | `repodata_from_packages.json` file | application/vnd.conda.repodata_from_packages.v1+json |
| repodata_from_packages.json.zst | `repodata_from_packages.json.zst` file | application/vnd.conda.repodata_from_packages.v1+json+zst |
| repodata_from_packages.json.gz | `repodata_from_packages.json.gz` file | application/vnd.conda.repodata_from_packages.v1+json+gzip |

> | repodata_from_packages.json.bz2 | `repodata_from_packages.json.bz2` file | application/vnd.conda.repodata_from_packages.v1+json+bz2 |

Reviewer: These follow the same schema as other repodatas, but are unpatched. Do we need a separate media type? What about `current_repodata`?

Author: Ah good point. I'll condense to a single repodata schema and add current repodata. Unfortunately current repodata starts with a "c" and so conflicts with a package named _repdata.json. We'll have to resolve that conflict somehow.

Reviewer: Argh, do we need to prefix something to all `repodata`s then? Because in that case we should prepend the same prefix to all non-package information for consistency. Something like `m`. So `repodata.json` would be stored under `mrepodata.json`, and `patch_instructions.json` would be under `mpatch_instructions.json`?

Reviewer: Sure, that can work. It would just be odd because in a way we are limiting `repodata.json`-like files to _not_ start with `h` or `c`, and at the same time allowing all the other letters implicitly, while also requiring some copies to start with `r`? So I'd rather have a clear and simple rule for non-package files, so they all must start with the same prefix.

Author: I see what you mean. That is indeed better so let's go with `m` for `metadata`. I will adjust the CEP.

into the OCI blobs and what additional metadata (if any) is stored in the OCI manifest.

A conda package as an OCI artifact MUST have the following layers:

- a layer holding either the `.tar.bz2` or `.conda` package file with the mediaType
  `application/vnd.conda.package.v1` or `application/vnd.conda.package.v2`, respectively.
- a layer holding the `info` folder as a gzipped tarball with extension `.tar.gz` and mediaType `application/vnd.conda.info.v1.tar+gzip`.
- a layer holding the `info/index.json` file as plain JSON with the mediaType `application/vnd.conda.info.index.v1+json`.

Additional layers are NOT allowed in the OCI artifact. If a conda package exists in both `.tar.bz2` and `.conda` format, the `.conda`
version MUST be chosen to be in the OCI artifact.

The manifest MUST have the following [OCI Annotations][annotations]

- `org.conda.oci.schema` with the value `1` indicating the `v1` specification (defined by this document).
- `org.conda.package.name` with the name of the conda package.
- `org.conda.package.version` with the version of the conda package.

> - `org.conda.package.build` with the build string of the conda package

Reviewer: Shall we add channel, label and subdir too?

Author: subdir yes, but not channel or label.

Author: We never hash the subdir so no need actually.

Additional annotations under the `org.conda` namespace are NOT allowed.

### OCI Artifact Structure for Repodata, etc

All artifacts in the following OCI repositories

```text
<OCI conda channel base URL>/<OCI-compatible subdir>/repodata.json
<OCI conda channel base URL>/<OCI-compatible subdir>/repodata_from_packages.json
<OCI conda channel base URL>/<OCI-compatible subdir>/run_exports.json
<OCI conda channel base URL>/<OCI-compatible subdir>/patch_instructions.json
```

MUST have a layer with the data as JSON and the appropriate mediaType from the table above. Additional layers with
`gzip`, `bz2`, or `zst` compressed copies of the data MAY be included.

The manifest MUST have the following [OCI Annotation][annotations]:

> - `org.conda.oci.schema` with the value `1` indicating the `v1` specification (defined by this document).

Reviewer: Same here: shall we annotate channel, subdir and label in case they end up being hashed?

Author: They would not be hashed. See above.

Author: ah sorry. these entries are never hashed so no need here.

  substantial feedback from the community. The current version of the specification is a result of that feedback. Particularly
  notable feedback included the suggestion of
  [symbol name mangling techniques and the first version of the shortened encoding used above](https://github.com/conda/ceps/pull/70/files#discussion_r1740438239)
  by [Bas Zalmstra](https://github.com/baszalmstra).
- The [channel-mirrors project](https://github.com/channel-mirrors) with its associated package
  [conda-oci-mirror](https://github.com/channel-mirrors/conda-oci-mirror) developed a proof-of-concept implementation of
  conda packages and repodata stored in an OCI registry. This specification is based on the experience gained from that project, but is not
  fully backwards compatible with the original implementation. See the [Backwards Compatibility](#backwards-compatibility) section above.
  This original implementation and implied specification is referred to as `v0`.
- One alternative to excluding certain labels and channels from distribution via an OCI conda channel is to define a further custom

>   encoding of the channel path parts of the URL. For example, one could use an encoding inspired by percent-encoding for URLS. This encoding

Reviewer: ```suggestion
  encoding of the channel path parts of the URL. For example, one could use an encoding inspired by percent-encoding for URLs. This encoding
```
